File: modules/reliquary/legacy/python/src/reliquary/hashing.py
"""Streaming checksums used for manifests and verify passes."""

from __future__ import annotations

import hashlib
from pathlib import Path

ALGORITHMS = ("sha256", "sha512")
CHUNK = 1024 * 1024


def hash_file(path: Path) -> dict[str, str]:
    sha256 = hashlib.sha256()
    sha512 = hashlib.sha512()
    with path.open("rb") as fh:
        while True:
            buf = fh.read(CHUNK)
            if not buf:
                break
            sha256.update(buf)
            sha512.update(buf)
    return {"sha256": sha256.hexdigest(), "sha512": sha512.hexdigest()}


def write_sum_files(directory: Path, files: list[Path]) -> None:
    lines256: list[str] = []
    lines512: list[str] = []
    for f in files:
        digests = hash_file(f)
        rel = f.name
        lines256.append(f"{digests['sha256']}  {rel}\n")
        lines512.append(f"{digests['sha512']}  {rel}\n")
    (directory / "SHA256SUMS").write_text("".join(lines256), encoding="utf-8")
    (directory / "SHA512SUMS").write_text("".join(lines512), encoding="utf-8")


def verify_sum_file(directory: Path, name: str = "SHA256SUMS") -> list[str]:
    """Return a list of problem strings. Empty means all good."""
    sums = directory / name
    if not sums.exists():
        return [f"missing {name}"]
    algo = "sha256" if "256" in name else "sha512"
    problems: list[str] = []
    for line in sums.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        digest, _, filename = line.partition("  ")
        if not filename:
            digest, _, filename = line.partition(" *")
        target = directory / filename
        if not target.exists():
            problems.append(f"{filename}: missing")
            continue
        got = hash_file(target)[algo]
        if got != digest:
            problems.append(f"{filename}: {algo} mismatch")
    return problems

File: modules/reliquary/src/hashing.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Read, Write};
use std::path::Path;

use sha2::{Digest, Sha256, Sha512};

use crate::error::Result;

const CHUNK: usize = 1024 * 1024;

#[derive(Debug, Clone)]
pub struct Digests {
    pub sha256: String,
    pub sha512: String,
}

pub fn hash_file(path: &Path) -> Result<Digests> {
    let mut sha256 = Sha256::new();
    let mut sha512 = Sha512::new();
    let mut f = File::open(path)?;
    let mut buf = vec![0u8; CHUNK];
    loop {
        let n = f.read(&mut buf)?;
        if n == 0 {
            break;
        }
        sha256.update(&buf[..n]);
        sha512.update(&buf[..n]);
    }
    Ok(Digests {
        sha256: hex::encode(sha256.finalize()),
        sha512: hex::encode(sha512.finalize()),
    })
}

pub fn write_sum_files(directory: &Path, files: &[std::path::PathBuf]) -> Result<()> {
    let mut lines256 = String::new();
    let mut lines512 = String::new();
    for f in files {
        let d = hash_file(f)?;
        let name = f.file_name().unwrap().to_string_lossy();
        lines256.push_str(&format!("{}  {}\n", d.sha256, name));
        lines512.push_str(&format!("{}  {}\n", d.sha512, name));
    }
    std::fs::write(directory.join("SHA256SUMS"), lines256)?;
    std::fs::write(directory.join("SHA512SUMS"), lines512)?;
    Ok(())
}

pub fn verify_sum_file(directory: &Path, name: &str) -> Result<Vec<String>> {
    let sums = directory.join(name);
    if !sums.exists() {
        return Ok(vec![format!("missing {name}")]);
    }
    let algo256 = name.contains("256");
    let mut problems = Vec::new();
    let f = File::open(&sums)?;
    for line in BufReader::new(f).lines() {
        let line = line?;
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let (digest, filename) = if let Some((d, n)) = line.split_once("  ") {
            (d, n)
        } else if let Some((d, n)) = line.split_once(" *") {
            (d, n)
        } else {
            continue;
        };
        // The sums file may have arrived from an untrusted medium (usb::pull_block
        // copies a whole block directory in before anything is verified), so the
        // filename is attacker data. write_sum_files only ever emits bare
        // file_name()s, so refusing anything with a separator costs nothing and
        // stops `<hash>  ../../../etc/shadow` turning verify into a read/existence
        // oracle over the whole filesystem. Unix-only crate, so `/` is the only
        // separator to worry about.
        if filename.is_empty() || filename.contains('/') || filename == ".." || filename == "." {
            problems.push(format!("{filename}: invalid filename in sums file"));
            continue;
        }
        let target = directory.join(filename);
        if !target.exists() {
            problems.push(format!("{filename}: missing"));
            continue;
        }
        let got = hash_file(&target)?;
        let actual = if algo256 { &got.sha256 } else { &got.sha512 };
        if actual != digest {
            problems.push(format!(
                "{filename}: {} mismatch",
                if algo256 { "sha256" } else { "sha512" }
            ));
        }
    }
    Ok(problems)
}

#[allow(dead_code)]
pub fn write_text(path: &Path, text: &str) -> Result<()> {
    let mut f = File::create(path)?;
    f.write_all(text.as_bytes())?;
    Ok(())
}
